iterator.go

```diff
@@ -176,27 +177,42 @@ func (it *Iterator) AllForPrefixes(prefixes [][]byte, limit int32, lastKey []byt
 				return nil, nil, err
 			}
 		} else {
-			ok = true
+			ok = preseekOk
 		}
 
 		acc := make([][]byte, 0, 1)
 
-		for ok && bytes.HasPrefix(it.key, prefix) && (values < limit) {
+		for ok && bytes.HasPrefix(it.key, prefix) {
 			last = it.key
 			acc = append(acc, it.Value())
-			values++
+
 			if ok, err = it.Next(); err != nil {
 				return nil, nil, err
 			}
+			values++
 
 			if !ok || !bytes.Equal(last, it.key) {
+				if values >= limit {
+					ok = false
+				}
 				cp := make([]byte, len(last))
 				copy(cp, last)
 				res[string(cp)] = acc
 				acc = make([][]byte, 0, 1)
 			}
 		}
+
+		if values < limit {
```

Comment:
This seems acceptable to be now. Trying to handle the case where values = limit and there are no more matching keys seems not worth the complexity.